levelup

tamasori opened the pull request:
@oliverosz Készen vagyok a levelup funkcióval, a teszteket is beraktam, és sikeresen lefutottak.

tamasori committed:
add Player class and header, inherited from Entity

tamasori committed:
change main properties to float in Entity

tamasori committed:
add experience logic and level logic to Player

tamasori committed:
add maxHealt getter, and change main properties and attributes to float

tamasori committed:
changed an Entity to Player in main

tamasori committed:
added Player equals Entity operator=

tamasori committed:
change an Entity to Player in Game class

tamasori committed:
removed ExperienceAtFightStart property from Player

tamasori committed:
change in-out test parameters

tamasori committed:
change GHA checkout to levelup branch

tamasori committed:
add cppcheck to GHA

tamasori committed:
CPPCHECK log

tamasori committed:
exclude information type from cppcheck

tamasori committed:
Merge branch 'levelup' of https://github.com/Teaching-projects/SZE-MOSZE-2020-Pottyok into levelup

tamasori committed:
Fail test if cppcheck finds error or warning

tamasori committed:
CPPCHECK log

tamasori committed:
add memory check

tamasori committed:
add Valgrind

tamasori committed:
fail test on Valgrind memory leak

tamasori committed:
fix Valgrind test

oliverosz commented on .github/workflows/test.yml:
```diff
@@ -5,7 +5,17 @@ jobs:
     steps:
       - uses: actions/checkout@v2
+        with:
+          ref: levelup
```
Ez nem kell ide, sőt, nem jó, ha itt van. Ha be lesz mergelve a masterbe, akkor a későbbi branchekre nem fog így lefutni.

oliverosz commented on .github/workflows/test.yml:
```diff
@@ -5,7 +5,17 @@ jobs:
+          github_username: "tamasori"
+          github_email: "oeri.tamas@gmail.com"
+          commit_msg: "CPPCHECK log"
+          enable: "style,warning,performance,portability,unusedFunction,missingInclude"
```
Feladat:

>A GHA workflow hasaljon el, ha barmilyen memoriaszemet keletkezik, vagy a statikus kodellenorzo barmilyen errort / warningot talal. A stilus/teljesitmeny uzeneteket csak tegye bele egy fajlba, es artifactolja el.

Sajnos ez a deep5050-féle Action nem jó erre a célra. Egyrészt a workflow simán továbbmegy pipával, bármilyen hibát talál is a cppcheck, másrészt a reportot becommitolja a repóba, ahelyett, hogy artifactolná. Ez nem jó, mert így keveredik a kódokkal, és beleszemetel a commit historyba is.

oliverosz commented on .github/workflows/test.yml:
```diff
@@ -5,7 +5,17 @@ jobs:
+          ref: levelup
       - name: Build
-        run: g++ -std=c++17 *.cpp -o output
+        run: g++ -fsanitize=leak -fsanitize=address -std=c++17 *.cpp -o output
```
A warningok is legyenek bekapcsolva, legalább a `-Wall`, és ezeket tekintse fordítási hibának.

oliverosz commented on Entity.cpp:
```diff
@@ -4,11 +4,10 @@
 #include <regex> 
 
-Entity::Entity(const int health,const int attackDamage,const std::string name){
+Entity::Entity(const float health,const float attackDamage,const std::string name) : Name(name){
```
Az egyszerű paramétereknél felesleges a const, a string viszont legyen const referenciaként átadva. És az összes adattag inicializáló listával legyenek inicializálva, ne csak a név.

oliverosz commented on Player.h:
```diff
@@ -0,0 +1,23 @@
+class Player : public Entity{
+    protected:
+        float ExperienceCurrent;
+        float ExperienceAtLastRound;
```
Ez miért tartozik hozzá a Player aktuális állapotához?

tamasori replied:
Elnézést, de itt nem értem mi a probléma. A tapasztalati pontokhoz kapcsolódik a kérdés?

oliverosz replied:
Az `ExperienceAtLastRound`-ot nem értem, hogy miért tartozik hozzá az objektum állapotához. Ha csak egy támadásnál van rá szükség, a szintlépés kiszámolásához, akkor elég ott egy lokális változó erre.

tamasori replied:
Ja már értem, jogos, javítom.

oliverosz commented on main.cpp:
```diff
@@ -16,13 +17,13 @@ int isNumeric(const string & chars) {
-    Entity ent1(150, 10, "Maple");
-    Entity ent2(45, 30, "Sally");
+    Player ent1(800, 150, "Kakarott");
+    Entity ent2(450, 300, "Sally");
```
A másik félnek is Playernek kéne lennie. Mindkét harcos fejlődjön a csata közben.

oliverosz commented on tests.sh:
```diff
@@ -15,6 +18,14 @@ done < "$expected"
 
 different="$(diff $expected $results)"
 
+if grep -q "(warning)" "./cppcheck_report.txt"; then
+  exit 1
+fi
+
+if grep -q "(error)" "./cppcheck_report.txt"; then
+  exit 1 
+fi
+
```
Ja látom, itt van ellenőrizve, és emiatt a test lépés el fog hasalni, ha a cppcheck errort/warningot talál. De továbbra se jó, hogy be van commitolva a report. Meg logikailag sem jó az, hogy a cppcheck és az in-out test lépések keverednek. Ezek egymástól független dolgok.

tamasori replied:
Csináljak neki egy külön workflowt, vagy elég, ha csinálok egy másik test.sh fájlt?

oliverosz replied:
Nem muszáj külön workflow, de valahogy legyen elszeparálva külön lépésbe.

tamasori committed:
add loose gained levels and reset stats to start of fight on loose

tamasori committed:
CPPCHECK log

oliverosz commented on Player.cpp:
```diff
@@ -0,0 +1,52 @@
+
+Player::Player(const int health,const int attackDamage,const std::string name) : Entity(health, attackDamage, name) {
+    this->ExperienceAtLastRound         = 0;
+    this->ExperienceCurrent             = 0;
```
Inicializáló lista itt is.

tamasori committed:
Revert "add loose gained levels and reset stats to start of fight on loose"

This reverts commit 13d7ba6d796cd407438a6f864e5c4b7ece1ccde5.

tamasori committed:
CPPCHECK log

tamasori committed:
change Player2 to Player from Entity

tamasori committed:
fix GitHub actions, separate checks

tamasori committed:
fix two runs error in github actions

tamasori committed:
fix permission error

tamasori committed:
remove cpp_check.sh replaced by inline code

tamasori committed:
Create copy constructors, and make code cleaner

tamasori committed:
add Wno-reorder

tamasori committed:
add test run before valgrind

tamasori committed:
fix valgrind fail only in GHA(try)

tamasori committed:
fix valgrind

tamasori committed:
remove ExperienceAtLastRound, replaced by local variable

tamasori commented:
@oliverosz  elvégeztem a kért javításokat

oliverosz commented on .github/workflows/test.yml:
```diff
@@ -5,7 +5,20 @@ jobs:
     steps:
       - uses: actions/checkout@v2
+      - name: Install dependencies
+        run: sudo apt-get install -y valgrind && sudo apt-get install -y cppcheck
```
(OK, de amúgy az apt-getnek fel lehet sorolni több csomagot is szóközzel elválasztva)

oliverosz commented on .github/workflows/test.yml:
```diff
@@ -5,7 +5,20 @@ jobs:
+      - name: Install dependencies
+        run: sudo apt-get install -y valgrind && sudo apt-get install -y cppcheck
+      - name: CPPCheck Static Code Analysis
+        run: cppcheck --enable=all ./*.cpp 2> cppcheck_report.txt && if grep -q "(warning)" "./cppcheck_report.txt"; then exit 1; fi && if grep -q "(error)" "./cppcheck_report.txt"; then exit 1; fi
```
Ez se rossz, de amúgy a cppchecknek van egy `--error-exitcode` kapcsolója is. :D Többet ésszel, mint erővel!

oliverosz commented on .github/workflows/test.yml:
```diff
@@ -5,7 +5,20 @@ jobs:
-        run: g++ -std=c++17 *.cpp -o output
-      - name: Test
+        run: g++ -Wall -Werror -Wno-reorder -std=c++17 *.cpp -o output
+      - name: Test run after build
+        run: ./output
+      - name: Valgrind memory check
+        run: valgrind ./output
+      - name: In-Out tests
```
Az in-out teszteknél úgyis le van tesztelve a futás, ezért a "Test run after build" lépés felesleges. A memchecknél viszont ha van leak, akkor ne kapjon pipát a workflow.

oliverosz commented on .github/workflows/test.yml:
```diff
@@ -5,7 +5,20 @@ jobs:
       - name: Build
-        run: g++ -std=c++17 *.cpp -o output
-      - name: Test
+        run: g++ -Wall -Werror -Wno-reorder -std=c++17 *.cpp -o output
```
A jelzett hibát kellene kijavítani, nem a warningot kikapcsolni. A rossz sorrendben történő inicializálás valóban okozhat hibákat, ez hasznos figyelmeztetés.

oliverosz commented on Player.cpp:
```diff
@@ -0,0 +1,54 @@
+Player::Player(const int health,const int attackDamage,const std::string& name) : ExperienceCurrent(0), Entity(health, attackDamage, name) {}
+
+Player::Player(const Player& other) : Entity(other.Health, other.AttackDamage, other.Name){
+    this->MaxHealth                     =   other.MaxHealth;
+    this->Health                        =   other.Health;
+    this->AttackDamage                  =   other.AttackDamage;
+    this->Name                          =   other.Name;
+    this->ExperienceCurrent             =   other.ExperienceCurrent;
```
Az első 4-et beállítja az Entity, itt már nem kell. Az utolsó meg legyen inicializáló listában.

oliverosz commented on Player.cpp:
```diff
@@ -0,0 +1,54 @@
+}
+
+Player& Player::operator=(const Player &player){
+    this->MaxHealth                     =   player.MaxHealth;
+    this->Health                        =   player.Health;
+    this->AttackDamage                  =   player.AttackDamage;
+    this->Name                          =   player.Name;
```
Itt is lehetne ezekhez használni az `Entity::operator=`-t. Meg egy `dynamic_cast`tal össze lehetne vonni ezt a két assignment operátort.

tamasori committed:
remove unnecessary test run and set valgrind --error-exitcode to 1

tamasori committed:
remove -Wno-reorder, fix initialization list order

tamasori committed:
using one function for operator =, implementing dynamic_cast

tamasori committed:
remove empty row in expected

tamasori committed:
display more info in in-out test

tamasori committed:
file contents display

tamasori committed:
fix in-out tests, remove dc

tamasori committed:
more info in cppcheck
on my computer, cppcheck does not find any warnings, or errors, thats why I am using GHA to test.

tamasori committed:
remove cc from Player

tamasori commented:
@oliverosz Megcsináltam a változatásokat, a dupla operator= törölve lett, mivel rájöttem, hogy volt 1-2 felesleges függvény, és azok a függvények miatt csináltam, hogy a cppcheck ne hasaljon el. 

Plusz észrevettem, hogy GHA-n és a gépemen is a legfrissebb CPPCheck van telepítve, mégis mást eredmény adnak ugyan azon fájlokra. A GHA többször elhasalt, pedig nálam csak style és performance megjegyzéseket dobott. A GHA-n mások az alap beállítások? Vagy valamiben különbözik? Papíron ugyanúgy kéne működnie mégis mást eredményt adnak.

oliverosz reviewed (approved):
Nem tudom, hogy ugyanaz a cppcheck verzió hogy adhat más kimenetet azonos beállításokkal azonos bemenetre. Látom, a dynamic_castos assignment operátort nem sikerült megoldani, de így is elfogadható, meg a workflow is rendben van már.

oliverosz commented:
@hegyhati merge

hegyhati commented:
@oliverosz adtam jogot.
